### ISSUE TITLE
Problem with shaders on retina display

### ISSUE THREAD

**Issue by bolso80** [cocos2d/cocos2d-x#17046]
We are trying to draw a circle with shaders and use some math functions to calculate the distance from the center of the circle to the coordinates of pixel to draw.

We do something like this:

First attempt (just use distance function)

```
varying vec4 v_fragmentColor;
varying vec2 v_texCoord;

uniform vec2 position;
uniform float lightSize;
uniform float darkness;

void main()
{
    gl_FragColor = vec4(vec3(0,0,0), 1.0);
    
    float distanceValue = distance(gl_FragCoord.xy, position);
    
    if (distanceValue < lightSize) {
        gl_FragColor = vec4(vec3(1,0,0), 1.0);
    }
}
```

This code does a mask with a red circle without opacity from the "position" like the center of the cirlcle. 

This works fine on non retina displays, but when we tried it on retina displays, it works for small values, but when we started to increase them (we noticed it to stop working around 300) it behaves as if for every pixel the condition is true.

Next we do this test:

Second Attempt: Replace distance function

```
varying vec4 v_fragmentColor;
varying vec2 v_texCoord;

uniform vec2 position;
uniform float lightSize;
uniform float darkness;

void main()
{
    gl_FragColor = vec4(vec3(0,0,0), 1.0);
    
    vec2 distanceVector = gl_FragCoord.xy - position.xy;
    float distanceValue = sqrt(pow(distanceVector.x, 2.0) + pow(distanceVector.y, 2.0));
    
    if (distanceValue < lightSize) {
        gl_FragColor = vec4(vec3(1,0,0), 1.0);
    }
}
```

We tried doing the distance calculation by hand and see that when the distanceVector has negative components it was always returning as if the if condition was true (so we got the top left, bottom left, and bottom right quads as red and in the top right corner we got the right behaviour with a quarter red circle and the rest of the quad masked.

NON RETINA DISPLAY:

![screen shot 2016-12-20 at 4 09 00 pm](https://cloud.githubusercontent.com/assets/6196495/21364177/79b95e00-c6ce-11e6-961e-31c0d9949e5b.png)

RETINA DISPLAY:

![img_0056](https://cloud.githubusercontent.com/assets/6196495/21364192/89394f5c-c6ce-11e6-898b-e0f60cef705a.png)

Third attempt: use manual distance but making distanceVector component possitive.

```
varying vec4 v_fragmentColor;
varying vec2 v_texCoord;

uniform vec2 position;
uniform float lightSize;
uniform float darkness;

void main()
{
    gl_FragColor = vec4(vec3(0,0,0), 1.0);
    
    vec2 distanceVector = abs(gl_FragCoord.xy - position.xy);
    float distanceValue = sqrt(pow(distanceVector.x, 2.0) + pow(distanceVector.y, 2.0));
    
    if (distanceValue < lightSize) {
        gl_FragColor = vec4(vec3(1,0,0), 1.0);
    }
}
```

We thought we had solved the issue with this (we don't know why it failed originally though) because on iOS 8 it worked, but when we tried on iOS 9 it failed in the same way as the original code.

Maybe we are doing something wrong in our shader calculation, but we think it is weird to get different behavior on different devices.

**Comment by minggo** [comment]
Could you please provide a test case?

**Comment by minggo** [comment]
I think the problem may be:

* gl_FragCoord is resolution depend
* position is always in pixel as cocos2d-x enable retina by default

**Comment by bolso80** [comment]
Hi @minggo 

I copy below a link with 3 examples on file fragmentShader.frag, calculating the distance with "distance" function, calculating the distance manualy without abs, and another with abs.

On ipad retina with iOS 9 all the examples behave incorrectly, but on iOS 8 the last one works fine. 
On non retina display all of them work fine (when testing on mac they also work).

https://drive.google.com/open?id=0B88Hp_AZaft0ZHJqbGZpaVVJYnc

The example is using cocos2d-x version 3.13.1

Thanks

**Comment by minggo** [comment]
I tested it, it works fine on Mac OS X(10.12.2). And it is wrong on my iPhone6(10.1.1), the whole screen will be red at the end. I tested example 1, i will test others and try to find the reason.

Edit: i used latest codes in v3.14 branch.

**Comment by minggo** [comment]
I have not idea of this issue. @ricardoquesada any idea?

**Comment by ricardoquesada** [comment]
@minggo I think it is as you said... we are not using independent resolution variables in our shaders.

Is this a v3.14 bug? otherwise I can take a look at it for v3.15

**Comment by minggo** [comment]
No, it is not a v3.14 bug.

**Comment by pablorealini** [comment]
Do you have any viable workaround for this or maybe we should wait for a fix in 3.15?
Thanks both!

**Comment by minggo** [comment]
It it caused by precision. You can added `precision highp float;` at the begin of fragmentShader.frag.  

@ricardoquesada i think we can change the default precision to highp, i met similar issue a few days ago.
